Source/CBLMultipartBuffer.m

```diff
@@ -0,0 +1,79 @@
+//
+//  CBLMultipartBuffer.m
+//  CouchbaseLite
+//
+//  Created by Robert Payne on 1/03/16.
+//  Copyright © 2016 Couchbase, Inc. All rights reserved.
+//
+
+#import "CBLMultipartBuffer.h"
+
+@interface CBLMultipartBuffer() {
+    NSMutableData *_data;
+    NSUInteger _offset;
+}
+
+@end
+@implementation CBLMultipartBuffer
+
+@synthesize compactionLength=_compactionLength;
+
+#pragma mark - Initialization
+
+- (instancetype)init {
+    if((self = [super init])) {
+        _data = [NSMutableData dataWithCapacity:1024];
+        _offset = 0;
+        _compactionLength = 4096;
+    }
+    return self;
+}
+
+#pragma mark - Actions
+
+- (BOOL)advance:(NSUInteger)amount {
+    if (_offset + amount >= _data.length) {
+        Warn(@"Preventing unsafe buffer overflow", nil);
+        return NO;
+    }
+    _offset += amount;
+    return YES;
+}
+- (BOOL)hasBytesAvailable {
+    return _data.length > _offset;
+}
+- (NSUInteger)bytesAvailable {
+    if(_data.length > _offset) {
+        return _data.length - _offset;
+    }
+    return 0;
+}
+- (void)appendData:(NSData *)data {
+    [_data appendData:data];
+}
+- (void)appendBytes:(const void *)bytes length:(NSUInteger)length {
+    [_data appendBytes:bytes length:length];
+}
+- (void)compact {
+    if(_offset > _compactionLength && _offset > (_data.length >> 1)) {
+        _data = [NSMutableData dataWithBytes:(char *)_data.bytes + _offset
```

Comment:
I don't think the optimisation is worth changing it here, it's a minute optimisation for how often the compaction runs.

The reduction of the capacity is actually intentional to ensure memory stays as compact as possible.